### PR TITLE
Changed importError to ModuleNotFoundError

### LINKED ISSUE #11523: importorskip variant that uses ModuleNotFoundError
#### What's the problem this feature will solve?
sometimes I install a package eg PyTables and it fails to compile properly and fails with an ImportError with an undefined symbol, however I don't see this in CI because I only do `pytest.importorskip("pytables")`

#### Describe the solution you'd like
a new kwarg to importorskip or a new function that catches ModuleNotFoundError instead of ImportError eg
`importorskip(exc=ModuleNotFoundError)` or `findmoduleorskip(...)`

#### Alternative Solutions
N/A

#### Additional context
Originally asked on discord

### PR DESCRIPTION
**Pull request by shekhuverma**
Closes #11523

Changed ImportError to ModuleNotFoundError in importorskip()``

### CODE REVIEW

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @shekhuverma.

Can you please add a test that shows `importorskip` will only skip on an explicit `ModuleNotFoundError`, but produce a failure for an `ImportError` exception?

File: testing/test_runner.py
Comment (nicoddemus) [inline testing/test_runner.py:786]: Instead of mocking `builtins.__import__`, which seems dangerous, let's instead use `pytester` and generate actual modules and try to import them;

```suggestion
def test_importorskip_module_not_found_error(pytester: Pytester) -> None:
    """
    importorskip() will only skip modules if they were not actually found via ``ModuleNotFoundError``, instead of catching the more generic ``ImportError``.
    
    #11523
    """
    fn = pytester.makepyfile("raise ImportError('some specific problem')")
    pytester.syspathinsert()
    
    with pytest.raises(ImportError, match="some specific problem"):
        pytest.importorskip(fn.stem)
        
    # User might decide to skip on any ImportError.    
    pytest.importorskip(fn.stem, exc=ImportError)  # passes
```

File: testing/test_runner.py
Comment (The-Compiler) [inline testing/test_runner.py:770]: Please use [monkeypatch](https://docs.pytest.org/en/stable/how-to/monkeypatch.html) instead of hand-rolling this.

File: src/_pytest/outcomes.py
Comment (nicoddemus) [inline src/_pytest/outcomes.py:228]: Let's add a new parameter to this function, `*, exc_type: Type[ImportError] = ModuleNotFoundError` and use it here:

```suggestion
        except exc_type as exc:
```

This way we provide a escape hatch to users which rely on the old behavior.

Also please update the docs, describing the parameters and explicitly showing an example for users that want the old behavior.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
I think at the very least, we need an easy escape hatch to get the former behavior, no? I'm somewhat ambivalent on adding a warning (after all, it's a loud failure - the only difference would be a warning vs. failing collection).

However, if someone really wants to catch `ImportError` here, they should be able to do `pytest.importorskip("mymodule", exc=ImportError)` or somesuch (with `exc=`, we might want to ensure that `issubclass(exc, ImportError)` maybe, at least on the type annotation level).

Right now, if someone wants to keep the existing behavior, they'd need to go from `pytest.importorskip("mymodule")` to something like (untested):

```python
try:
    import mymodule
except ImportError:
    pytest.skip("some good reason here", allow_module_level=True)
```

or so, which seems quite burdensome.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Besides the changes requested, please also add a new changelog entry, as improvement.

I actually think we probably should not introduce the new behavior right away, instead capture `ImportError` and show a warning stating that the default behavior will change in future versions to catch `ModuleNotFoundError` instead.

**Comment by nicoddemus** [comment]
@The-Compiler and I left similar reviews almost at the same time hehehe.

I think we should actually not change the default behavior right away, for better or worse this is a breaking change.

I think introducing a new parameter `exc_type: Optional[Type[ImportError]] = None`, that when `None`:

1. We capture `ModuleNotFoundError` first; in that case skip.
2. We capture `ImportError` next, in that case issue a warning and also skip.

In 9.0, we turn the warning into an error, and 9.1 remove (2) from the code and let the `ImportError` propagate unchanged.

The good thing is that users can use `exc_type` today in code, will not break anything that passes today, and provides a escape hatch for users that don't care and want to skip on any `ImportError`.

**Comment by shekhuverma** [comment]
Thanks for the review.
I made the requested changes, let me know if there is something else to be done

File: src/_pytest/outcomes.py
Comment (nicoddemus) [inline src/_pytest/outcomes.py:235]: Sorry I was not clearer.

I meant that we should only issue the warning if:

1. The user is not passing an explicit `exc_type`.
2. We actually capture `ImportError`.

So something like:


```python
    with warnings.catch_warnings():
        # Make sure to ignore ImportWarnings that might happen because
        # of existing directories with the same name we're trying to
        # import but without a __init__.py file.
        warnings.simplefilter("ignore")

        if exc_type is None:
            exc_type = ImportError
            warn_on_import_error = True
        else:
            warn_on_import_error = False 

        try:
            __import__(modname)
        except exc_type as exc:
            if reason is None:
                reason = f"could not import {modname!r}: {exc}"
            if warn_on_import_error and type(exc) is ImportError:
                warnings.warn(
                    PytestDeprecationWarning(
                        f"pytest.importorskip() caught {exc}, but this will change in a future pytest release to only capture ModuleNotFoundError exceptions by default.\nTo overwrite the future behavior and silence this warning, pass exc_type=ImportError explicitly.",
                    )
                )
            raise Skipped(reason, allow_module_level=True) from None
```

So we will not be changing the behavior right now, only in the future.

Do you agree @The-Compiler?

Comment (The-Compiler) [reply src/_pytest/outcomes.py:235]: Yep, that sounds like a solid approach. We'll also need tests for the warning and the different argument behaviours.

Comment (shekhuverma) [reply src/_pytest/outcomes.py:235]:
```python
   def importorskip(
    modname: str,
    minversion: Optional[str] = None,
    reason: Optional[str] = None,
    exc_type: Optional[Type[ImportError,ModuleNotFoundError]] = None,
    ) -> Any:
    .
    .
    .
    if exc_type is None:
       #to keep the old behavior
       exc_type = ImportError
       warn_on_import_error = True
    else:
        warn_on_import_error = False
        
    if warn_on_import_error or type(exc) is ImportError:
        warnings.warn()
```
I think the function should be like this, showing a warning if the user doesn't pass any exc_type or when there is ImportError

Comment (nicoddemus) [reply src/_pytest/outcomes.py:235]: We do not want to issue the warning unless we actually caught `ImportError`.

Most of the time the exception being raised is `ModuleNotFoundError`, which is OK and what users expect. 

The problem is when the library is found, but `ImportError` is raised because of a related problem (for example failure to load a required a shared library). In this case we want to issue a warning now, and in pytest 9.0, let the exception propagate upwards.

The `exc_type` parameter will let users be explicit about what they want to capture.

**Comment by nicoddemus** [comment]
@shekhuverma thanks for following up!

Took the liberty of adding more tests, docs, and change the implementation a bit.

@The-Compiler would appreciate another review.

**Review by nicoddemus — APPROVED** [review]

**Comment by nicoddemus** [comment]
After this gets merged I will create an issue to track changing the default in `9.1`.

**Comment by shekhuverma** [comment]
Thanks for the help @The-Compiler and @nicoddemus.
This was my first open-source contribution.
Please mark more issues that I can solve

**Comment by shekhuverma** [comment]
Hello there, what are the next steps?
when will it get merged to the main?

**Comment by nicoddemus** [comment]
The ball is on our side now @shekhuverma, waiting on @The-Compiler's review.

**Comment by nicoddemus** [comment]
(Note to ourselves, we need to squash merge this)

**Comment by The-Compiler** [comment]
Sorry for the delay! I plan to look at this today, now that PyConDE is over.

File: doc/en/deprecations.rst
Comment (The-Compiler) [inline doc/en/deprecations.rst:40]: nitpick:

```suggestion
Catching only :class:`ModuleNotFoundError` by default (and letting other errors propagate) would be the best solution,
```

File: src/_pytest/outcomes.py
Comment (The-Compiler) [inline src/_pytest/outcomes.py:282]: Why not call `warning.warn(...)` here directly, given that `warning = ...` is only set from here?

File: src/_pytest/outcomes.py
Comment (The-Compiler) [inline src/_pytest/outcomes.py:287]: Same here, couldn't this be simplified by moving the `raise Skipped(...)` to directly after the warning?

File: src/_pytest/outcomes.py
Comment (The-Compiler) [inline src/_pytest/outcomes.py:272]: How do we want to handle custom subclasses of `ImportError`? Wouldn't a

```suggestion
            if warn_on_import_error and not issubclass(exc, ModuleNotFoundError):
```

be more appropriate here?

File: testing/test_runner.py
Comment (The-Compiler) [inline testing/test_runner.py:793]: I love this! Never seen it before, I've always done:

```python
assert len(captured) == 1
warning = captured[0]
```

for this kind of thing.

File: src/_pytest/outcomes.py
Comment (The-Compiler) [inline src/_pytest/outcomes.py:251]: I'm a bit confused by the Codecov annotations here. It looks to me like this (and the others) should be covered by tests just fine? Are the added tests not running or something? Or is codecov broken? Or am I just missing something?

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks @shekhuverma and @nicoddemus. I have a couple of remaining questions and comments (most of them minor, but one semantic thing that I think we should aim to get correct).

File: src/_pytest/outcomes.py
Comment (nicoddemus) [reply src/_pytest/outcomes.py]: Hmm good point. If some library/user out there is raising an `ImportError` subclass (but not subclassing `ModuleNotFoundError`) we probably want to warn as well about the incoming behavior change, good catch.

File: src/_pytest/outcomes.py
Comment (nicoddemus) [reply src/_pytest/outcomes.py:282]: Because at this point we are inside the `with warnings.catch_warnings():` block.

File: testing/test_runner.py
Comment (nicoddemus) [reply testing/test_runner.py:793]: All credits go to https://www.cosmicpython.com, seen this pattern there and have been using it ever since. 😁

File: src/_pytest/outcomes.py
Comment (nicoddemus) [reply src/_pytest/outcomes.py:251]: Not sure, but this is definitely being covered by tests (and normal tests, not through `pytester` either). I even added a hard `assert 0` in this line to ensure it was hit by the tests, and it is (as expected).

Not sure what's going on.

File: src/_pytest/outcomes.py
Comment (nicoddemus) [reply src/_pytest/outcomes.py:287]: As per https://github.com/pytest-dev/pytest/pull/12220#discussion_r1580025393, we need to raise this after the warning. 👍

**Comment by shekhuverma** [comment]
Can you please mark more related issues/improvements that I can try?
I tried searching in the issues section but failed to choose one.
Thanks

**Review by The-Compiler — APPROVED** [review]

**Comment by The-Compiler** [comment]
Thanks again! :+1: We have the [good first issue](https://github.com/pytest-dev/pytest/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) label, though indeed I think we haven't really been doing a good job at applying it to issues recently...

**Comment by shekhuverma** [comment]
> Thanks again! 👍 We have the [good first issue](https://github.com/pytest-dev/pytest/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) label, though indeed I think we haven't really been doing a good job at applying it to issues recently...

Yes, there is only one issue with that label now

**Pull request merged**
